ENH allow for grouped documentation with CommandCollection

tomMoral opened the pull request:
While building a CLI with many different subcommands, I splitted them in groups and merged together using a `CommandCollections`.
When documenting this CLI, I wanted to preserve this group structure. I ended up tweaking a bit the `sphinx-click` extension to show the groups in `sources` as sections, without breaking the API. See here for an example of the results obtained: https://517-227720375-gh.circle-artifacts.com/0/dev/cli.html#benchopt

I ported the code for this in this PR. Let me know if you think it might interest others, I can try to document this change and make a few examples if you think it is worth it :) Else, feel free to close this PR. It might be too specific to my use case.

stephenfin commented:
Oh, I didn't know this existed. Docs are [here](https://click.palletsprojects.com/en/7.x/api/#click.CommandCollection) for anyone interested. Given this is a native feature of click, I'd be happy to support this. Could you add some docs and tests to prove this out before I review it fully?

tomMoral commented:
I did a quick pass on the doc and the tests:

- I added an example and reference it in usage for the doc. I modified a bit the structure to make it more modular, let me know if you'd rather keep a flat structure, I will integrate the example directly in usage (but I feel it might be a bit heavy).
- I tested the formatter did not change on `nested=short` and is minimal on `nested=full` for a `CommandCollection`. I don't know how I could chack for the output of `_generate_nodes` which would probably be better to check for the tests.

stephenfin commented on docs/usage.rst:
```diff
@@ -147,6 +147,27 @@ the roles provided by the `Sphinx standard domain`_.
+group listed in the command ``sources``. The group names are used as section titles
+and the help string from the description are used as section description.
+Thus, a client defined using a |CommandCollection| as ``cli`` will be rendered
+using *sphinx-click* and the following directive
```
Can you add a colon at the end of this line

stephenfin commented on docs/usage.rst:
```diff
@@ -147,6 +147,27 @@ the roles provided by the `Sphinx standard domain`_.
+*sphinx-click* renders collection of commands with multiple sections, one for each
+group listed in the command ``sources``. The group names are used as section titles
+and the help string from the description are used as section description.
+Thus, a client defined using a |CommandCollection| as ``cli`` will be rendered
```
s/will be/can be/

stephenfin commented on docs/usage.rst:
```diff
@@ -147,6 +147,27 @@ the roles provided by the `Sphinx standard domain`_.
+      :prog: cli
+      :nested: full
+
+with the subcommands of each group in different sections, one for each group in
```
This will render the subcommands of...

stephenfin commented on docs/usage.rst:
```diff
@@ -147,6 +147,27 @@ the roles provided by the `Sphinx standard domain`_.
+      :nested: full
+
+with the subcommands of each group in different sections, one for each group in
+``sources``. An example is provided in :ref:`example_commandcollections`.
```
nit: Can you use a `:doc:` link instead?

stephenfin commented on docs/examples/commandcollections.rst:
```diff
@@ -0,0 +1,27 @@
+.. _example_commandcollections:
+
+Documentating |CommandCollection|
```
Documenting

stephenfin commented on docs/examples/commandcollections.rst:
```diff
@@ -0,0 +1,27 @@
+Documentating |CommandCollection|
+=================================
+
+The client in the file ``examples/commandcollections/cli.py`` using a
```
I don't think we need to reference the file when we're dumping the contents here. How about simply:

    Consider the following sample application, using |CommandCollection|:

stephenfin commented on docs/examples/commandcollections.rst:
```diff
@@ -0,0 +1,27 @@
+=================================
+
+The client in the file ``examples/commandcollections/cli.py`` using a
+|CommandCollection| such as
```
Can you add a colon

stephenfin commented on docs/examples/commandcollections.rst:
```diff
@@ -0,0 +1,27 @@
+
+.. literalinclude:: ../../examples/commandcollections/cli.py
+
+The automatic documentation using *sphinx-click* gives for the following directive:
```
This can be documented using *sphinx-click* like so:

stephenfin reviewed (changes requested):
> I did a quick pass on the doc and the tests:

Thanks!

> * I added an example and reference it in usage for the doc. I modified a bit the structure to make it more modular, let me know if you'd rather keep a flat structure, I will integrate the example directly in usage (but I feel it might be a bit heavy).

No issues from me. It's still referenced from the main `usage` doc, which is what I care about. We might expand these docs at some point, but the example doc is useless regardless.

>     * I tested the formatter did not change on `nested=short` and is minimal on `nested=full` for a `CommandCollection`. I don't know how I could chack for the output of `_generate_nodes` which would probably be better to check for the tests.

Yeah, I need to integrate the `sphinx.testing` module to add proper Sphinx tests, but haven't gotten around to doing so yet. I don't expect you to do this so what you have is good enough.

I have a couple of nits with the docs inline, but in general this is excellent. If you can address those, I'm happy to merge this and cut a new release.

stephenfin commented:
Thanks!

tomMoral commented:
Thanks a lot for the super fast review!
